pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -665,22 +948,107 @@ func (b *binaryOperationSeriesBuffer) close() {
 	}
 }
 
-type binaryOperationFunc func(left, right float64) float64
+type binaryOperationFunc func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error)
 
+// FIXME(jhesketh): Investigate avoiding copying histograms for binary ops.
 var arithmeticOperationFuncs = map[parser.ItemType]binaryOperationFunc{
-	parser.ADD: func(left, right float64) float64 {
-		return left + right
+	parser.ADD: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs != nil {
+			res, err := hlhs.Copy().Add(hrhs)
+			if err != nil {
+				return 0, nil, false, err
+			}
+			return 0, res.Compact(0), true, nil
+		}
+		return lhs + rhs, nil, true, nil
+	},
+	parser.SUB: func(lhs, rhs float64, hlhs, hrhs *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+		if hlhs != nil && hrhs != nil {
+			res, err := hlhs.Copy().Sub(hrhs)
+			if err != nil {
+				return 0, nil, false, err
+			}
+			return 0, res.Compact(0), true, nil
+		}
+		return lhs - rhs, nil, true, nil
+	},
```

Comment:
Are these correct if one side is a float and the other is a histogram?

Reply:
You can't add or subtract the different types as far as I understand. I think you'd get `histogram_sum` and then add to that if that's what you're trying to achieve.

These functions are lifted [verbatim from promql](https://github.com/prometheus/prometheus/blob/main/promql/engine.go#L2720).

Reply:
Right, so the expected behaviour in the case where we have a float and a histogram is to ignore the histogram and treat it as if it was 0?